world_positioning/src/worldP_ekf.py

```diff
@@ -0,0 +1,381 @@
+
+        '''
+
+        #first version: only 1 object in state
+
+        self.nState = 13 #number of elements in state
+
+        #n and m are only for 1 point
+        self.n = self.nState+3 #substate + world x,y,z of object
+        self.m = self.nState+3 #sub state + x,y,z of object in camera
+
+        self.x = np.zeros((self.n,1))
+        self.x[IDx.Ow] = 1
+
+        self.u = np.zeros((self.m, 1))
+
+        self.B = np.array() #input modification matrix
+
+        self.P = np.eye(self.n)
+        self.Q = np.eye(self.n) * 0.01 #uncert in dynamics model
+
+        self.camera_info = ZedCamListener() 
+        self.state_info = SubStateListener()
+
+        self.state_msg = SubState()
+        self.camera_msg = Camera() #need to implement this for camera
+        self.world_msg = WorldState()
+
+        self.rate = 20
+        self.rate_ctrl = rospy.Rate(self.rate)
+
+        self.calc_F = jacobian(self.f)
+
+        self.world_pub = rospy.Publisher("world state", WorldState, queue_size = 1)
+
+        self.last_prediction_t = self.get_t()
+
+    def get_t(self):
+        return rospy.Time.now().to_sec()
+
+    def prediction(self):
+        #jacobian of F
+        F = self.calc_F(self.x, self.u)
+        F = np.reshape(F,(self.n ,self.n))
+        Fx = F[0:self.n, 0:self.n]
+        print("Fx", Fx)
+
+        #update x and P
+        self.x = self.f(self.x, self.u)
+	self.P = np.linalg.multidot(Fx,self.P, np.transpose(Fx)) + self.Q
+
+    def Yaw(self, y):
+        return np.matrix([
+                [math.cos(y), -math.sin(y), 0],
+                [math.sin(y),  math.cos(y), 0],
+                [          0,            0, 1]
+                ])
+
+    def Pitch(self, p):
+        return np.matrix([
+                [ math.cos(p), 0, math.sin(p)],
+                [               0, 1,               0],
+                [-math.sin(p), 0, math.cos(p)]
+                ])
+
+    def Roll(self, r):
+        return np.matrix([
+                [1,              0,               0],
+                [0, math.cos(r), -math.sin(r)],
+                [0, math.sin(r),  math.cos(r)]
+                ])
+
+    def RPY_Matrix(self, r, p, y):
+        r = Roll(r)
+        p = Pitch(p)
+        y = Yaw(y)
+        return np.linalg.multi_dot([y, p, r])
+    def update(self):
+        #update world based on sub state
+        z = np.zeros((0,0)) #measurements
+        h = np.zeros((0,0)) #measurement prediction
+
+        H = np.zeros((0,0)) #jacobian of h
+        R = np.zeros((0,0)) #uncert of measurement
+
+        def add_block_diag(H,newH):
+            if H.shape[0] == 0:
+                ans = np.diag(newH)
+                ans.shape = (newH.shape[0],newH.shape[0])
+                return ans
+            return scipy.linalg.block_diag(H,newH)
+
+        def add_block_vert(H,newH):
+            if H.shape[0] == 0:
+                return newH
+            return np.vstack([H,newH])
+
+        def read_listener_state(listener, z, h, H, R):
```

Comment:
I'm a bit confused about what would happen if i removed the point rotation data processing in read_listener_camera. Won't `z=np.append(z,np.array([meas]) ` load the wrong data onto z? The `state_to_measurement_h` functoin in the listener class seems to be only putting the correct values into the `pred`, not the `z`. Is there something I'm not understanding, like the `z` in the` state_to_measurement_h` function in worldP_listener.py is the same `z` in the worldP_ekf.py file?

Reply:
This can be a choice of which data we want to compare. The choice may have some effects on the quality of the estimation, but I don't think we're in a place where we can fully measure that. Beyond that, there's no 'wrong' form of the data, just inconsistent formats. We can have z and h/pred both be absolute coordinates, or both be relative coordinates.

The incoming data is the relative position of the object, and we also have estimates of the object an sub's current positions. Those can be used to calculate the relative position when creating ```pred```. I think using the relative measurement form will be the best way. If z is modified, then it can be affected by inaccuracies in the state, whereas the prediction of the measurement would be supposed to have this inaccuracy anyway.

It also means that we can use one function for all of the listeners, and there are now less places where a change could mean something goes wrong.

Reply:
Just to clarify, I think the two scenarios we could use would be:

1)
z = raw sensor reading = relative position
pred = calculated relative position from state vector
now we compare relative positions

2)
z = raw sensor reading, transformed by current estimate from state vector
pred = 'raw' elements of the state vector
now we compare absolute positions


Each option has one 'raw' part, and one that is calculated using more information. 2 is what I believe you were aiming at, but I think 1 may be the best option.